pkg/probe/probe.go

```diff
@@ -114,6 +128,102 @@ func (p *prober) probe(ctx context.Context, signals chan Signal) error {
 	return nil
 }
 
+func (p *prober) sendProxyProtocolHeaders(ctx context.Context,
+	ppMode ProxyProtocolMode,
+	localIp, remoteIp, localPortStr, remotePortStr string,
+) error {
+	if ppMode == ProxyProtocolDisabled {
+		return fmt.Errorf("proxy protocol is disabled")
+	}
+
+	log := util.CtxLogOrPanic(ctx).With(
+		zap.String("ppVersion", string(ppMode)),
+	)
+
+	localPort, err := strconv.Atoi(localPortStr)
+	if err != nil {
+		return fmt.Errorf("failed to convert local port to int: %w", err)
+	}
+	remotePort, err := strconv.Atoi(remotePortStr)
+	if err != nil {
+		return fmt.Errorf("failed to convert remote port to int: %w", err)
+	}
+
+	header := &proxyproto.Header{
+		Version:           1,
+		Command:           proxyproto.PROXY,
+		TransportProtocol: proxyproto.TCPv4,
+		SourceAddr: &net.TCPAddr{
+			IP:   net.ParseIP(localIp),
+			Port: localPort,
+		},
+		DestinationAddr: &net.TCPAddr{
+			IP:   net.ParseIP(remoteIp),
+			Port: remotePort,
+		},
+	}
+
+	if ppMode == ProxyProtocolV2 {
+		header.Version = 2
+	}
+	log = log.With(zap.Any("header", header))
+
+	// serialize the header
+	ppBytes, errFormat := header.Format()
+	if errFormat != nil {
+		return fmt.Errorf("failed to serialize proxy protocol %s header: %w", ppMode, errFormat)
+	}
+
+	// ppv1 is human readable
+	if ppMode == ProxyProtocolV1 {
+		log = log.With(zap.String("raw-v1-header", string(ppBytes)))
+	}
+
+	log.Info("sending proxy protocol headers")
+	_, err = bytes.NewBuffer(ppBytes).WriteTo(p.conn)
+	if err != nil {
+		return fmt.Errorf("failed to write proxy protocol %s headers: %w", ppMode, err)
+	}
+	log.Info("sent proxy protocol headers")
+	return nil
+}
+
+func (p *prober) maybeSendProxyProtocolHeaders(ctx context.Context, signals chan Signal) {
```

Comment:
Nit: `sendProxyHeadersIfEnabled`/`sendProxyHeadersIfRequired`? Slightly more context at a glance

Reply:
I'm ok with "maybe", because we use it in similar way in other places, and at this point it's an established pattern

Reply:
Also: it is a mixed set of conditions: it needs to be "requested/enabled" by the user, and also a connection must've been successfully established. That would form an awkward name when trying to be precise.

Thanks folks!